lib/SplatArgPass.cpp

```diff
@@ -152,32 +153,31 @@ bool SplatArgPass::runOnModule(Module &M) {
         uint32_t vec_size = param_info.vector_size;
         bool last_is_scalar = func_info.getLastParameter().vector_size == 0;
         if (vec_size != 0 && last_is_scalar) {
-          bool has_3_params =
-              func_type == Builtins::kClamp || func_type == Builtins::kMix;
-          std::string NewFName =
-              getSplatName(func_info, param_info, has_3_params);
-          Function *NewCallee = getReplacementFunction(F, NewFName);
-          // Replace the users of the function.
-          for (User *U : F.users()) {
-            replaceCall(NewCallee, dyn_cast<CallInst>(U));
-          }
-          func_list.push_front(&F);
+          func_list.push_back({&F, func_info});
         }
         break;
       }
-      default:
-        break;
       }
     }
   }
-  if (func_list.size() != 0) {
-    // remove dead
-    for (auto *F : func_list) {
-      if (F->use_empty()) {
-        F->eraseFromParent();
-      }
+  // Replace with vectorized version
+  for (auto FI : func_list) {
+    auto *F = FI.first;
+    auto &func_info = FI.second;
+    auto func_type = func_info.getType();
+    auto &param_info = func_info.getParameter(0);
+    bool has_3_params =
+        func_type == Builtins::kClamp || func_type == Builtins::kMix;
+    std::string NewFName = getSplatName(func_info, param_info, has_3_params);
+    Function *NewCallee = getReplacementFunction(*F, NewFName);
+    // Replace the users of the function.
+    for (User *U : F->users()) {
+      replaceCall(NewCallee, dyn_cast<CallInst>(U));
```

Comment:
This is still modifying the user list while iterating over it, so my test case still hangs with this code.

Reply:
Thanks for catching that.

@sjw36 I'd suggest using a SmallVector to copy the users before modifying them.

Reply:
Too right, I will fix that..

Reply:
I was just going to :
  while (!F->use_empty()) {
    User *U = *F->users().begin();
    replaceCall(...);
  }
Good?

Reply:
Looks ok, but I don't have time to try it out.